interfaces/cython/cantera/yaml2ck.py

```diff
@@ -0,0 +1,1047 @@
+
+    :param val:
+        Float to be formatted into a str
+    :param precision:
+        How many digits after the decimal
+    :param exp_digits:
+        How many digits the exponent has
+    """
+    if math.isinf(val) and val > 0:
+        return " inf"
+
+    elif math.isinf(val):
+        return "-inf"
+
+    # round to correct number of significant digits
+    if val != 0:
+        val = round(val, precision - int(math.floor(math.log10(abs(val)))) - 1)
+
+    s = f"{val: .{precision}e}"
+
+    mantissa, exponent = s.split("e")
+    exp_digits += 1  # +1 due to sign
+
+    return f"{mantissa}E{int(exponent):+0{exp_digits}d}"
+
+
+def build_header_text(solution, max_width=120):
+    """
+    Creates the header text for all output files
+
+    :param solution:
+        The `solution` object being converted
+    :param max_width:
+        The maximum width of lines before they start to wrap
+    """
+    text = []
+
+    if "description" in solution.input_header:
+        text_wrap = lambda t: fill(
+            t,
+            width=max_width,
+            break_long_words=False,
+            break_on_hyphens=False,
+            initial_indent="! ",
+            subsequent_indent="!   ",
+        )
+
+        text = solution.input_header["description"].splitlines()
+        text = [text_wrap(line) for line in text]
+    else:
+        note = ""
+
+    ct_note = (
+        f"! Chemkin file was converted from a Cantera {ct.__version__} "\
+        "`solution` object"
+    )
+
+    has_ct_note = False
+    for n, line in enumerate(text):
+        if "! Chemkin file was converted from a Cantera" in line:
+            text[n] = ct_note
+            has_ct_note = True
+            break
+
+    if not has_ct_note:
+        text.append(f"!\n{ct_note}\n! \n\n")
+
+    return "\n".join(text)
+
+
+def build_species_text(solution, max_width=80, sort_elements=True, 
+                       sort_species=True):
+    """
+    Creates element and species declarations
+
+    :param solution:
+        The `solution` object being converted
+    :param max_width:
+        The maximum width of lines before they start to wrap
+    :param sort_elements:
+        A boolean flag to indicate if elements should be sorted
+    :param sort_species:
+        A boolean flag to indicate if species should be sorted
+    """
+
+    def build_elements(elements, atomic_weights):
+        if sort_elements:  # Sort elements by atomic weight
+            elements = [c for _, c in sorted(zip(atomic_weights, elements))]
+
+            if sort_elements:
+                elements = elements[::-1]
+
+        elements_text = fill(
+            "  ".join(elements),
+            width=max_width,
+            break_long_words=False,
+            break_on_hyphens=False,
+        )
+
+        text = ["ELEMENTS"]
+        text.append(f"{elements_text}")
+        text.append("END\n\n")
+
+        return "\n".join(text)
+
+    def build_species(solution_species):
```

Comment:
Why is `atomic_weights` an argument to `build_elements` above, but `molecular_weights` isn't an argument here? I'd suggest being consistent.

But actually, I think a better method overall would be to move `build_species()` and `build_elements()` outside this function, take a `Solution` and the respective sorting argument and the max width as their arguments, and have this function simply call those two. I don't think they need to be nested.

Reply:
The elements section is unique because there isn't a `solution_elements` that I can pull from. I have to pull those properties from the Solution. I can't make the inputs consistent without cluttering up the code.

The reason I have them as functions defined within `build_species_text` is that they are only ever used together writing the mechanisms. Much of how I wrote these is basically making modules for the portions being written. If this is not good enough reason in your mind to keep them together then I can unnest them.